zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledCall.java

```diff
@@ -37,183 +38,181 @@
-      limitListener.onIgnore();
-      throw e;
+  /**
+   * To simplify code, this doesn't actually invoke the underlying {@link #execute()} method. This
+   * is ok because in almost all cases, doing so would imply invoking {@link #enqueue(Callback)}
+   * anyway.
+   */
+  @Override protected Void doExecute() throws IOException {
+    AwaitableCallback awaitableCallback = new AwaitableCallback();
+    doEnqueue(awaitableCallback);
+    if (!await(awaitableCallback.countDown)) throw new InterruptedIOException();
+
+    Throwable t = awaitableCallback.throwable.get();
+    if (t != null) {
+      if (t instanceof Error) throw (Error) t;
+      if (t instanceof IOException) throw (IOException) t;
+      if (t instanceof RuntimeException) throw (RuntimeException) t;
+      throw new RuntimeException(t);
     }
+    return null;
   }
 
-  @Override protected void doEnqueue(Callback<V> callback) {
-    Listener limitListener = limiter.acquire(null)
+  @Override protected void doEnqueue(Callback<Void> callback) {
+    Listener limiterListener = limiter.acquire(null)
       .orElseThrow(RejectedExecutionException::new); // TODO: make an exception message
+    LimiterReleasingCallback releasingCallback =
+      new LimiterReleasingCallback(callback, isOverCapacity, limiterListener);
 
     try {
-      executor.execute(new QueuedCall<>(this, callback, limitListener));
-    } catch (RuntimeException | Error e) {
-      propagateIfFatal(e);
-      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
-      // write bound, but a drop in concurrency won't necessarily help.
-      limitListener.onIgnore();
-      throw e;
+      executor.execute(new EnqueueAndAwait(this, releasingCallback));
+    } catch (RuntimeException | Error t) { // possibly rejected, but from the executor, not storage!
+      propagateIfFatal(t);
+      callback.onError(t);
+      // Ignoring in all cases here because storage itself isn't saying we need to throttle. Though
+      // we may still be write bound, but a drop in concurrency won't necessarily help.
+      limiterListener.onIgnore();
+      throw t;
     }
   }
 
-  @Override public Call<V> clone() {
-    return new ThrottledCall<>(executor, limiter, isOverCapacity, delegate.clone());
+  @Override public Call<Void> clone() {
+    return new ThrottledCall(delegate.clone(), executor, limiter, isOverCapacity);
   }
 
   @Override public String toString() {
     return "Throttled(" + delegate + ")";
   }
 
-  static String setCurrentThreadName(String name) {
-    Thread thread = Thread.currentThread();
-    String originalName = thread.getName();
-    thread.setName(name);
-    return originalName;
+  static final class AwaitableCallback implements Callback<Void> {
+    final CountDownLatch countDown = new CountDownLatch(1);
+    final AtomicReference<Throwable> throwable = new AtomicReference<>();
```

Comment:
I'm not immediately seeing the need for this to be an AtomicReference.  You create a new instance per-Call and only one Thread should execute that Call.  So you can save some time/space by just leaving this as a normal reference.

Reply:
ah yeah. I should add a comment. true that only one thread will set this. but the problem is that the thread invoking the callback isn't the same as the thread reading the exception. we could consider volatile though I suppose. would that be better?

Reply:
You can consider `volatile` though I personally wouldn't.  I believe the semantics around Threads and variable assignment are safe enough that the assignment of that value should resolve before attempting to read it.  Simply because the Thread doing the read is waiting for the entire writing Thread to finish before progressing.  So the write implictly has to succeed before the read will even be attempted.  Assuming of course that this is only used in `doExecute()`.

Normally, `volatile` variables are used where there is a high likelihood of near-concurrent read/write.  Which again, I don't think is the case here.

Reply:
You would generally need `volatile` when writes and reads happen in different threads just to make sure the writes were already flushed to memory, it's less about the concurrency of the read/write.

But here it is actually clear that we don't need to use `volatile` since `CountDownLatch` guarantees writes are flushed before `await()` returns

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html
`Memory consistency effects: Until the count reaches zero, actions in a thread prior to calling countDown() happen-before actions following a successful return from a corresponding await() in another thread.`